tooling/nargo_cli/src/cli/test_cmd.rs

```diff
@@ -68,258 +94,410 @@ pub(crate) fn run(args: TestCommand, config: NargoConfig) -> Result<(), CliError
+                            name: test.name,
+                            package_name: test.package_name,
+                            status,
+                            output,
+                            time_to_run,
+                        };
+
+                        // It's fine to ignore the result of sending.
+                        // If the receiver has hung up, everything will wind down soon anyway.
+                        let _ = thread_sender.send(test_result);
+                    })
+                    .unwrap();
             }
-            FunctionNameMatch::Contains(pattern) => {
-                return Err(CliError::Generic(format!("Found 0 tests containing '{pattern}'.",)))
+
+            // Also drop main sender so the channel closes
+            drop(sender);
+
+            // We'll go package by package, but we might get test results from packages ahead of us.
+            // We'll buffer those here and show them all at once when we get to those packages.
+            let mut buffer: HashMap<String, Vec<TestResult>> = HashMap::new();
+            for (package_name, test_count) in test_count_per_package {
+                let mut test_report = Vec::new();
+
+                // How many tests are left to receive for this package
+                let mut remaining_test_count = *test_count;
+
+                // Check if we have buffered test results for this package
+                if let Some(buffered_tests) = buffer.remove(package_name) {
+                    remaining_test_count -= buffered_tests.len();
+
+                    for test_result in buffered_tests {
+                        self.display_test_result(&test_result)
+                            .expect("Could not display test status");
+                        test_report.push(test_result);
+                    }
+                }
+
+                if remaining_test_count > 0 {
+                    while let Ok(test_result) = receiver.recv() {
+                        // This is a test result from a different package: buffer it.
+                        if &test_result.package_name != package_name {
+                            buffer
+                                .entry(test_result.package_name.clone())
+                                .or_default()
+                                .push(test_result);
+                            continue;
+                        }
+
+                        self.display_test_result(&test_result)
+                            .expect("Could not display test status");
+                        test_report.push(test_result);
+                        remaining_test_count -= 1;
+                        if remaining_test_count == 0 {
+                            break;
+                        }
+                    }
+                }
+
+                display_test_report(package_name, &test_report)
+                    .expect("Could not display test report");
+                test_reports.extend(test_report);
```

Comment:
Good point. I just kept it like that because it was like that, but I think we only need to know if there are any tests results, and if there are failures. I'll change it to track those things.

Reply:
Cool, thanks. It's only because that way we don't have to think about whether the reporting has already happened: it must have, because we can't do it on the return value. Also frees up the output strings.

Reply:
Done!